### ISSUE TITLE
asyncio worker threads are prematurely shut-down on SIGINT

### ISSUE THREAD

**Issue by uSpike** [agronholm/anyio#310]
```python
import anyio
import time

def thread():
    print("thread start", time.time())
    time.sleep(5)
    print("thread end", time.time())

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(anyio.to_thread.run_sync, thread)
        try:
            await anyio.sleep_forever()
        finally:
            await anyio.to_thread.run_sync(thread)

anyio.run(main, backend="asyncio")
```
If i run this and press CTRL-C while sleeping, it does this:

```
thread start 1623770620.0201666
^Cthread end 1623770625.0241108
thread start 1623770625.0242603
thread end 1623770630.028106
```

with `backend="trio"` I get what I expect:
```
thread start 1623770659.6473012
^Cthread start 1623770660.5315428
thread end 1623770664.652128
thread end 1623770665.5361295
```

One workaround for this issue is to use a signal receiver:
```python
import anyio
import time
import signal

def thread():
    print("thread start", time.time())
    time.sleep(5)
    print("thread end", time.time())

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(anyio.to_thread.run_sync, thread)
        try:
            with anyio.open_signal_receiver(signal.SIGINT) as signals:
                async for signum in signals:
                    break
        finally:
            await anyio.to_thread.run_sync(thread)

anyio.run(main, backend="asyncio")
```

Another workaround (which I'm using) is to spawn threads manually (`threading.Thread(...).start()`) for tear-down.  Both of these are asyncio-specific workarounds, and I think that anyio `to_thread.run_sync` should behave the same if running on asyncio or trio, if possible.

My use case is this is working with something that is synchronous and must use worker threads.  I must call the appropriate `close()` methods on the synchronous API from a thread so as to not block my async application.

**Comment by agronholm** [comment]
What happens here is that when the `await` is cancelled, the thread is put back on the list of idle threads, even though it's not idle yet. The next call to `run_sync()` then sees that idle thread and assigns a new task to it, instead of spawning a new worker thread.

**Comment by agronholm** [comment]
We might need to revert back to having the thread add itself to the list of idle workers. I need to ensure that this won't cause any regressions.

**Comment by uSpike** [comment]
Using #319 and I have another interesting case, maybe.

Raising `KeyboardInterrupt` in the main task works as expected where `thread()` is executed twice in parallel and run to completion.
```python
import anyio
import functools
import time

def thread():
    print("thread start", time.time())
    time.sleep(1)
    print("thread end", time.time())

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(functools.partial(anyio.to_thread.run_sync, thread))
        try:
            await anyio.sleep(0.1)
            raise KeyboardInterrupt()
        finally:
            await anyio.to_thread.run_sync(thread)

anyio.run(main, backend="asyncio")
```

However, raise `KeyboardInterrupt` in a task and it never runs `thread()` a second time, instead it hangs forever.
```python
import anyio
import functools
import time

def thread():
    print("thread start", time.time())
    time.sleep(1)
    print("thread end", time.time())

async def raise_soon():
    await anyio.sleep(0.1)
    raise KeyboardInterrupt()

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(functools.partial(anyio.to_thread.run_sync, thread))
        tg.start_soon(raise_soon)
        try:
            await anyio.sleep(0.5)
        finally:
            await anyio.to_thread.run_sync(thread)

anyio.run(main, backend="asyncio")
```
```
thread start 1623950796.8734636
Traceback (most recent call last):
  File "/home/jspeicher/src/xtf/test.py", line 23, in <module>
    anyio.run(main, backend="asyncio")
  File "/home/jspeicher/src/xtf/venv/lib/python3.9/site-packages/anyio/_core/_eventloop.py", line 56, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/home/jspeicher/src/xtf/venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 230, in run
    return native_run(wrapper(), debug=debug)
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/runners.py", line 47, in run
    _cancel_all_tasks(loop)
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/runners.py", line 63, in _cancel_all_tasks
    loop.run_until_complete(
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/jspeicher/src/xtf/venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 225, in wrapper
    return await func(*args)
  File "/home/jspeicher/src/xtf/test.py", line 21, in main
    await anyio.to_thread.run_sync(thread)
  File "/home/jspeicher/src/xtf/venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 564, in __aexit__
    raise exceptions[0]
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/jspeicher/src/xtf/test.py", line 12, in raise_soon
    raise KeyboardInterrupt()
KeyboardInterrupt
thread end 1623950797.8745818
^CException ignored in: <module 'threading' from '/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/threading.py'>
Traceback (most recent call last):
  File "/home/jspeicher/.pyenv/versions/3.9.2/lib/python3.9/threading.py", line 1428, in _shutdown
    lock.acquire()
KeyboardInterrupt: 

Command terminated
```
